qcutilities.py

```diff
@@ -284,6 +333,137 @@ def readBackCheck(rootTree, dict_Names, device, gtx, vt1bump=0):
+    vt1bump - value that has been added to CFG_THR_ARM_DAC
+    """
+
+    import numpy as np
+    import root_numpy as rp
+    import os, sys
+
+    # Check that the requested register is supported
+    list_KnownRegs = [
+            "CFG_PULSE_STRETCH",
+            "CFG_SYNC_LEVEL_MODE",
+            "CFG_SELF_TRIGGER_MODE",
+            "CFG_DDR_TRIGGER_MODE",
+            "CFG_SPZS_SUMMARY_ONLY",
+            "CFG_SPZS_MAX_PARTITIONS",
+            "CFG_SPZS_ENABLE",
+            "CFG_SZP_ENABLE",
+            "CFG_SZD_ENABLE",
+            "CFG_TIME_TAG",
+            "CFG_EC_BYTES",
+            "CFG_BC_BYTES",
+            "CFG_FP_FE",
+            "CFG_RES_PRE",
+            "CFG_CAP_PRE",
+            "CFG_PT",
+            "CFG_EN_HYST",
+            "CFG_SEL_POL",
+            "CFG_FORCE_EN_ZCC",
+            "CFG_FORCE_TH",
+            "CFG_SEL_COMP_MODE",
+            "CFG_VREF_ADC",
+            "CFG_MON_GAIN",
+            "CFG_MONITOR_SELECT",
+            "CFG_IREF",
+            "CFG_THR_ZCC_DAC",
+            "CFG_THR_ARM_DAC",
+            "CFG_HYST",
+            "CFG_LATENCY",
+            "CFG_CAL_SEL_POL",
+            "CFG_CAL_PHI",
+            "CFG_CAL_EXT",
+            "CFG_CAL_DAC",
+            "CFG_CAL_MODE",
+            "CFG_CAL_FS",
+            "CFG_CAL_DUR",
+            "CFG_BIAS_CFD_DAC_2",
+            "CFG_BIAS_CFD_DAC_1",
+            "CFG_BIAS_PRE_I_BSF",
+            "CFG_BIAS_PRE_I_BIT",
+            "CFG_BIAS_PRE_I_BLCC",
+            "CFG_BIAS_PRE_VREF",
+            "CFG_BIAS_SH_I_BFCAS",
+            "CFG_BIAS_SH_I_BDIFF",
+            "CFG_BIAS_SH_I_BFAMP",
+            "CFG_BIAS_SD_I_BDIFF",
+            "CFG_BIAS_SD_I_BSF",
+            "CFG_BIAS_SD_I_BFCAS",
+            "CFG_RUN",
+            "HW_CHIP_ID",
+            "VFAT_CHANNELS.CHANNEL"]
+
+    for regName in dict_Names.values():
+        if regName not in list_KnownRegs:
+            print "readBackCheckV3() does not understand %s"%(regName)
+            print "readBackCheckV3() is only supported for registers:", list_KnownRegs
+            sys.exit(os.EX_USAGE)
+
+    # Get data from tree
+    list_bNames = dict_Names.keys()
+    list_bNames.append('vfatN')
+    if "mask" in dict_Names.keys() or "trimDAC" in dict_Names.keys() or "trimPolarity" in dict_Names.keys():
+        list_bNames.append('vfatCH')
+    array_writeVals = rp.tree2array(tree=rootTree, branches=list_bNames)
+
+    # Get data from VFATs
+    for bName,regName in dict_Names.iteritems():
+        print "Reading back (%s,%s) from all VFATs, any potential mismatches will be reported below"%(bName,regName)
+        print "="*40
+
+        regMap = []
+        regValues = []
+        if regName == "VFAT_CHANNELS.CHANNEL": #Channel Register
+            regValues = getChannelRegisters(vfatBoard,mask)
+            for vfat in range(0,24):
+                if (mask >> vfat) & 0x1:
```

Comment:
should we check here if the mask was applied correctly?

Reply:
I'm not sure I understand what you are referring to.

In L428 `mask` comes from the user supplied `vfatmask` and is not a writeable register; the channel mask is bit number 15 of the integers stored in `regValues` (L426) and is checked below at L435.

Reply:
we can provide a check to see whether this VFAT is in a sleep mode (if accessible) or inaccessible. So that we dont have unexpected extra chip in run mode with garbage values in its settings

Reply:
I'm not sure I understand.  This function can check the currently stored register values against some externally supplied values.  

The list of supported registers is defined in `list_KnownRegs`.  So some other script should call this function (`readBackCheckV3`) for say `CFG_RUN` or some other register of interest (provided a `TTree` with the stored values of `rootTree` is provided).  But to add such a check here is now how the function was designed to be used.

Right now we do not have a `TTree` which defines say `CFG_RUN`.

Reply:
Again the goal here with this PR is to continue to bring the existing `v2b` functionality to the `v3` tools.